src/main/java/uk/gov/hmcts/reform/blobrouter/services/storage/RejectedFilesHandler.java

```diff
@@ -0,0 +1,103 @@
+package uk.gov.hmcts.reform.blobrouter.services.storage;
+
+import com.azure.storage.blob.BlobClient;
+import com.azure.storage.blob.BlobContainerClient;
+import com.azure.storage.blob.BlobServiceClient;
+import org.slf4j.Logger;
+import uk.gov.hmcts.reform.blobrouter.data.EnvelopeRepository;
+import uk.gov.hmcts.reform.blobrouter.data.model.Envelope;
+import uk.gov.hmcts.reform.blobrouter.data.model.Status;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.List;
+
+import static java.util.stream.Collectors.groupingBy;
+import static org.slf4j.LoggerFactory.getLogger;
+
+public class RejectedFilesHandler {
+
+    private static final Logger logger = getLogger(RejectedFilesHandler.class);
+
+    private static final String REJECTED_CONTAINER_SUFFIX = "-rejected";
+
+    private final BlobServiceClient storageClient;
+    private final EnvelopeRepository envelopeRepository;
+
+    public RejectedFilesHandler(BlobServiceClient storageClient, EnvelopeRepository envelopeRepository) {
+        this.storageClient = storageClient;
+        this.envelopeRepository = envelopeRepository;
+    }
+
+    /**
+     * Handles files that were rejected. Ie:
+     * - moves files to container for rejected files
+     * - removes files from the original container
+     * - marks envelopes in the DB as deleted
+     */
+    public void handle() {
+        List<Envelope> rejectedEnvelopes = envelopeRepository.find(Status.REJECTED, false);
+
+        logger.info("Found {} rejected envelopes", rejectedEnvelopes.size());
+
+        rejectedEnvelopes
+            .stream()
+            .collect(groupingBy(e -> e.container))
+            .forEach((container, envelopes) -> {
+                logger.info("Moving rejected files from container {}", container);
+
+                var sourceContainer = storageClient.getBlobContainerClient(container);
+                var targetContainer = storageClient.getBlobContainerClient(container + REJECTED_CONTAINER_SUFFIX);
+
+                envelopes.forEach(envelope -> handleEnvelope(sourceContainer, targetContainer, envelope));
+            });
+    }
```

Comment:
Maybe it would be worth logging the completion. Unless the task does it.

Reply:
done